### PR TITLE
TRY-JW-89902-Admin-Create-Group

### PR DESCRIPTION
**Pull request by JackWCDS**
[*] Groups now have a single owner
[+] GroupUserRoles table
 - clustered primary key of GroupUser_Id and GroupRole_Id
[-] Removed MembershipRole_Id from GroupUser table
- refactored to use GroupUsersInRoles table
[*] GetGroup now retrieves:- OwnerId, OwnerFirstName and OwnerSurname
- added new properties to Group record class
- updated GroupDataProvider SQL
[+] SeedData for GroupUser table
[+] Endpoint and service code for create group

.

### CODE REVIEW

**Comment by Matt-Beldon-CDS** [comment]
This seems to be breaking `RolesDataProvider.GetUserAndGroupUserRolesAsync` with the error message `Invalid column name 'MembershipRole_Id'`

**Review by Matt-Beldon-CDS — CHANGES_REQUESTED** [review]
This seems to be breaking `RolesDataProvider.GetUserAndGroupUserRolesAsync` with the error message `Invalid column name 'MembershipRole_Id'`

File: futurenhs.data/FutureNHS.Data.FutureNHS/dbo/Tables/GroupUsersInRoles.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿CREATE TABLE [dbo].[GroupUsersInRoles]
```
Comment (Matt-Beldon-CDS) [inline futurenhs.data/FutureNHS.Data.FutureNHS/dbo/Tables/GroupUsersInRoles.sql:1]: I think this table might need a rowversion too

Comment (Tim-Hunt-CDS) [reply futurenhs.data/FutureNHS.Data.FutureNHS/dbo/Tables/GroupUsersInRoles.sql:1]: Yes it will do

**Review by Tim-Hunt-CDS — COMMENTED** [review]
How do we deal with migrating data? Now it’s gone to production we can no longer lose any data so we need to migrate any existing user roles in groups to the new groupusersinroles table

**Comment by Matt-Beldon-CDS** [comment]
> 

Because we don't really use any migration scripts this is probably going to be a huge issue, not really sure how we're going to get around this. We can't be running scripts directly on prod after a deployment

**Comment by Tim-Hunt-CDS** [comment]
> > How do we deal with migrating data? Now it’s gone to production we can no longer lose any data so we need to migrate any existing user roles in groups to the new groupusersinroles table
> 
> Because we don't really use any migration scripts this is probably going to be a huge issue, not really sure how we're going to get around this. We can't be running scripts directly on prod after a deployment

Doesn’t feel like it was thought through before trying to deliver it on the last day of sprint.

**Comment by Tim-Hunt-CDS** [comment]
> > > How do we deal with migrating data? Now it’s gone to production we can no longer lose any data so we need to migrate any existing user roles in groups to the new groupusersinroles table
> > 
> > 
> > Because we don't really use any migration scripts this is probably going to be a huge issue, not really sure how we're going to get around this. We can't be running scripts directly on prod after a deployment
> 
> Doesn’t feel like it was thought through before trying to deliver it on the last day of sprint.

So what’s the solution? Because currently it’s blocking 2 stories we said we’d deliver this sprint

**Comment by Matt-Beldon-CDS** [comment]
> > Doesn’t feel like it was thought through before trying to deliver it on the last day of sprint.
> 
> So what’s the solution? Because currently it’s blocking 2 stories we said we’d deliver this sprint

I think this was always going to be an issue if we were having to make db changes once production went live, the main issue with this is that on deployment it will delete `MembershipRole_Id` from our group users table and then we'll lose that data. 

I think the only solution will be to keep the old columns, manually migrate the data and then at a later date we can delete the old columns

we're also probably going to have to make changes to the seed/automation data before we can properly test it

**Comment by Tim-Hunt-CDS** [comment]
> > > Doesn’t feel like it was thought through before trying to deliver it on the last day of sprint.
> > 
> > 
> > So what’s the solution? Because currently it’s blocking 2 stories we said we’d deliver this sprint
> 
> I think this was always going to be an issue if we were having to make db changes once production went live, the main issue with this is that on deployment it will delete `MembershipRole_Id` from our group users table and then we'll lose that data.
> 
> I think the only solution will be to keep the old columns, manually migrate the data and then at a later date we can go back and delete the old columns
> 
> we're also probably going to have to make changes to the seed/automation data before we can properly test it

Yep that’s what I was thinking. What I’m trying to point out is when you’re spotting areas for improvement, that’s great, but you need to fully think through their implementation and mitigate issues such as this, preferably not on the last day of sprint.

**Pull request closed without merging**